runtime/shared_common/CacheMap.cpp

```diff
@@ -444,143 +475,231 @@ SH_CacheMap::startup(J9VMThread* currentThread, J9SharedClassPreinitConfig* pico
-				 * to SH_CacheMap::startup():
-				 * 	- While marking cache full, last unused pages are also protected, which ideally should be done
-				 * 	  after protecting pages belonging to ROMClass area and metadata area.
-				 * 	- Secondly, when setting cache full flags, the code expects to be holding the write mutex, which is not done in
-				 * 	  SH_CompositeCacheImpl::startup().
-				 *
-				 * Do not call fillCacheIfNearlyFull() in readonly mode, as we cannot write anything to cache.
-				 */
-				if ((0 == (*_runtimeFlags & J9SHR_RUNTIMEFLAG_ENABLE_NESTED))
-					&& (!_ccHead->isRunningReadOnly())
-				) {
-					_ccHead->fillCacheIfNearlyFull(currentThread);
-				}
+		do {
+			++tryCntr;
+			if ((rc == CC_STARTUP_SOFT_RESET) && (deleteRC == -1)) {
+				/* If we've tried SOFT_RESET the first time around and the delete failed,
+			 	 * remove AUTOKILL so that we start up with the existing cache */
+				*runtimeFlags &= ~J9SHR_RUNTIMEFLAG_AUTOKILL_DIFF_BUILDID;
+			}
 
-				_ccHead->exitWriteMutex(currentThread, fnName);
+			rc = ccToUse->startup(currentThread, piconfig, cacheMemoryUT, runtimeFlags, _verboseFlags, _cacheName, cacheDirName, cacheDirPerm, &_actualSize, &_localCrashCntr, true, cacheHasIntegrity);
+			if (rc == CC_STARTUP_OK) {
+				if (sanityWalkROMClassSegment(currentThread, ccToUse) == 0) {
+					rc = CC_STARTUP_CORRUPT;
+					goto error;
+				}
 
-				if (CM_READ_CACHE_FAILED == itemsRead) {
-					Trc_SHR_CM_startup_Exit6(currentThread);
-					return -1;
+				if (!isCcHead) {
+					if (NULL == appendCacheDescriptorList(currentThread, _sharedClassConfig, ccToUse)) {
+						CACHEMAP_TRACE(J9SHR_VERBOSEFLAG_ENABLE_VERBOSE_DEFAULT, J9NLS_ERROR, J9NLS_SHRC_CM_FAILED_ALLOC_DESCRIPTOR);
+						Trc_SHR_CM_startup_Exit12(currentThread);
+						return -1;
+					}
 				}
-				if (CM_CACHE_CORRUPT == itemsRead) {
-					// TODO need a tracepoint here
-					rc = CC_STARTUP_CORRUPT;
+				if (ccToUse->enterWriteMutex(currentThread, false, fnName) == 0) {
+					UDATA idLen = 0;
+					bool isReadOnly = ccToUse->isRunningReadOnly();
+					if (false == isCcHead) {
+						if (strlen(cacheUniqueID) > 0) {
+							if (false == ccToUse->verifyCacheUniqueID(currentThread, cacheUniqueID)) {
+								/* modification to a low layer cache has been detected */
+								CACHEMAP_TRACE1(J9SHR_VERBOSEFLAG_ENABLE_VERBOSE_DEFAULT, J9NLS_ERROR, J9NLS_SHRC_CM_VERIFY_CACHE_ID_FAILED, cacheUniqueID);
+								ccToUse->exitWriteMutex(currentThread, fnName);
+								Trc_SHR_CM_startup_Exit8(currentThread);
+								return -1;
+							}
+						}
+					}
 
-					SH_Managers::ManagerWalkState state;
-					SH_Manager* walkManager = managers()->startDo(currentThread, 0, &state);
-					while (walkManager) {
-						walkManager->cleanup(currentThread);
-						walkManager = managers()->nextDo(&state);
+					if ((*runtimeFlags & J9SHR_RUNTIMEFLAG_ENABLE_REDUCE_STORE_CONTENTION) && !isReadOnly) {
```

Comment:
Please use `J9_ARE_ANY_BITS_SET` here and in similar situations.

Reply:
Fixed